### PR TITLE
UI control groups - detect control group

### PR DESCRIPTION
**Pull request by meirish**
This is the first step in supporting Control Groups (in Enterprise) in the UI. You can read more about Control Groups here: https://www.vaultproject.io/docs/enterprise/control-groups/index.html. 

At a high level you can gate access to any paths and specify a number of authorizations required, and what Identity Groups they're required from. This is done via a new block in ACL policies or via Sentinel with an EGP or RGP policy. If a user encounters a Control Group gated request, it will be a response wrapped token instead of the data returned from the path. 

In the UI we detect this via the global request path in the application adapter. If it's determined that the response _is_ a Control Group token, then we redirect to a new route where we'll instruct the user what they need to do and what the status of their token is. If you're using the console and request a control grouped path, we show an error in the console and link them to the route mentioned above.

The mechanism for this works as follows: 
There is a new control group service that contains the logic for detecting if a response is a control group. If a control group is not detected the method resolves the params to the adapter's ajax method. If it does detect one, the method rejects with the token information merged onto the error object so that Ember Data does not try to push the token data into the store. Where we make requests (the application adapter, and the console) then catch the error and either redirect or print an error.

Enterprise specific tests don't run in Travis, but the new test do all pass locally:

screen shot 2018-06-08 at 10 01 55 am

### CODE REVIEW

File: ui/app/services/control-group.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  router: inject.service(),
+
+  checkForControlGroup(callbackArgs, response, wasWrapTTLRequested) {
+    let creationPath = response && Ember.get(response, 'wrap_info.creation_path');
+    if (
+      this.get('version.isOSS') ||
+      wasWrapTTLRequested ||
+      !response ||
+      (creationPath && WRAPPED_RESPONSE_PATHS.includes(creationPath)) ||
+      (response && !response.wrap_info)
```
Comment (alisdair) [inline ui/app/services/control-group.js]: Since you're returning if `!response` is true above, I think you can simplify this like to `!response.wrap_info`.

File: ui/app/services/control-group.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    ) {
+      return RSVP.resolve(...callbackArgs);
+    }
+    let error = new ControlGroupError();
+    error = assign(error, response.wrap_info);
```
Comment (alisdair) [inline ui/app/services/control-group.js]: It's great that you have a custom error class. But is it necessary to flatten all attributes from `wrap_info` into the error? It might be clearer to be more explicit here and in the constructor. I had to dig around the tests to figure out what the expected attributes on `wrap_info` were.

File: ui/tests/acceptance/enterprise-control-groups-test.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  visit('/vault/secrets');
+  andThen(() => {
+    consoleComponent.runCommands([
+      'write sys/mounts/kv type=kv',
+      'write kv/foo bar=baz',
+      'write sys/auth/userpass type=userpass',
+      `write sys/policies/acl/kv-control-group policy='${POLICY}'`,
+      'write -field=client_token auth/token/create policies=kv-control-group'
+    ]);
+  });
```
Comment (alisdair) [inline ui/tests/acceptance/enterprise-control-groups-test.js]: 👨‍🍳 💋 👌

File: ui/tests/acceptance/tools-test.js [hunk 4]
```diff
@@ -7,9 +7,6 @@ moduleForAcceptance('Acceptance | tools', {
   beforeEach() {
     return authLogin();
   },
-  afterEach() {
-    return authLogout();
-  },
```
Comment (alisdair) [inline ui/tests/acceptance/tools-test.js]: Why this change? I don't follow.

File: ui/tests/pages/components/console/ui-panel.js [hunk 5]
```diff
@@ -1,10 +1,21 @@
...
   consoleInput: fillable('[data-test-component="console/command-input"] input'),
   consoleInputValue: value('[data-test-component="console/command-input"] input'),
   logOutput: text('[data-test-component="console/output-log"]'),
+  logOutputItems: collection('[data-test-component="console/output-log"] > div', {
+    text: text(),
+  }),
+  lastLogOutput: getter(function() {
+    let count = this.logOutputItems.length;
+    return this.logOutputItems.objectAt(count - 1).text;
+  }),
```
Comment (alisdair) [inline ui/tests/pages/components/console/ui-panel.js]: This is a nice touch.

File: ui/tests/unit/services/control-group-test.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    isEnt,
+    [[{'one': 'two', 'three': 'four'}], {foo: 'bar', wrap_info: {token: 'secret', accessor: 'lookup'}}, false],
+    (assert, result) => {
+      return result.then(
+        () => {},
+        (err) => {
+          assert.equal(err.token, 'secret')
+          assert.equal(err.accessor, 'lookup')
+        });
+    }
```
Comment (alisdair) [inline ui/tests/unit/services/control-group-test.js]: I think if the code (somehow) changed to `resolve` instead of `reject` for this case, the tests would still pass. Either an `assert.ok(false, 'should not resolve')` or an `assert.expect` in the test runner might help in that (unlikely) case.

File: ui/app/services/control-group.js (hunk 1 above)
Comment (meirish) [reply ui/app/services/control-group.js]: Oh yep, will change - also talked to the backend devs and this whole thing will be getting much simpler, just checking one flag once its in place.

File: ui/tests/acceptance/tools-test.js (hunk 4 above)
Comment (meirish) [reply ui/tests/acceptance/tools-test.js]: Oops, I thought I was seeing a race condition and took this out, but it was something else and it'd be better to have this in especially now that we're switching users.

File: ui/tests/unit/services/control-group-test.js (hunk 6 above)
Comment (meirish) [reply ui/tests/unit/services/control-group-test.js]: Good call! Added the assert.expect above.
